### PR TITLE
Fixes and improvements on StatefulWriter <2.0.x> [8324]

### PR DESCRIPTION
**Pull request by MiguelCompany**
This PR:
* Adds generic filtering mechanisms to the test_udpv4 transport
* Improves sending gaps for holes in history
* Fixes locator selection when re-sending samples to specific readers
* Fixes late-joining reliable volatile readers never notifying samples on some cases

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2059/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1815/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2053/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2061/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1817/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2055/)

File: test/blackbox/BlackboxTestsVolatile.cpp [hunk 1]
```diff
@@ -274,6 +276,205 @@ TEST_P(Volatile, AsyncVolatileSubBetweenPubs)
     reader.block_for_all();
 }
 
+TEST_P(Volatile, VolatileSubBetweenPubs)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/blackbox/BlackboxTestsVolatile.cpp:279]: We only specify `async` on the test names

File: test/blackbox/BlackboxTestsVolatile.cpp [hunk 2]
```diff
@@ -274,6 +276,205 @@ TEST_P(Volatile, AsyncVolatileSubBetweenPubs)
...
+    reader.startReception(data);
+    // Send data with some interval, to let async writer thread send samples
+    writer.send(data, 300);
+    // In this test all data should be sent.
+    ASSERT_TRUE(data.empty());
+    // Block reader until reception finished or timeout.
+    reader.block_for_all();
+}
+
+TEST_P(Volatile, AsyncVolatileSubBetweenTransientPubs)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/blackbox/BlackboxTestsVolatile.cpp:320]: The current name is in line with all the rest.

File: test/blackbox/BlackboxTestsVolatile.cpp [hunk 3]
```diff
@@ -274,6 +276,205 @@ TEST_P(Volatile, AsyncVolatileSubBetweenPubs)
...
+    reader.startReception(data);
+    // Send data with some interval, to let async writer thread send samples
+    writer.send(data, 300);
+    // In this test all data should be sent.
+    ASSERT_TRUE(data.empty());
+    // Block reader until reception finished or timeout.
+    reader.block_for_all();
+}
+
+TEST_P(Volatile, VolatileSubBetweenTransientPubs)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/blackbox/BlackboxTestsVolatile.cpp:363]: Same reasoning here

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2065/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1821/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2059/)

**Pull request merged**